### PR TITLE
Adding base vision package

### PR DESCRIPTION
**Pull request by ItayGabbay**
#### Reference Issues/PRs

Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests

#### Any other comments?

---
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/deepchecks/deepchecks/blob/main/CONTRIBUTING.rst

### CODE REVIEW

File: deepchecks/vision/checks/performance/performance_report.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+        Returns:
+            Self: instance of 'ClassPerformance' or it subtype
+
+        Raises:
+            DeepchecksValueError:
+                if unknown score function name were passed;
+        """
+        if score is None:
+            score = next(iter(MULTICLASS_SCORERS_NON_AVERAGE))
```
Comment (noamzbr) [inline deepchecks/vision/checks/performance/performance_report.py:218]: should probably be next on get_scorers_list. Maybe should save this result in an internal member?

File: deepchecks/vision/dataset.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            num_classes = len(samples_per_class.keys())
+            self._num_classes = num_classes
+        return self._num_classes
+
+    def get_samples_per_class(self):
+        if self._samples_per_class is None:
+            if self.label_type == TaskType.CLASSIFICATION.value:
+                counter = Counter()
+                for _ in range(len(self._data)):
+                    counter.update(self.label_transformer(next(iter(self._data))[1]))
```
Comment (noamzbr) [inline deepchecks/vision/dataset.py:58]: If anyone has a good idea of how to make sure that next(iter(self._data))[1] always gets self.label_transformer, it will be a welcome addition.

File: deepchecks/vision/dataset.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        return copy(self._samples_per_class)
+
+    def extract_label(self):
+        y = []
+        for _ in range(len(self._data)):
+            y.append(self.label_transformer(next(iter(self._data))[1]))
+        return cat(y, 0)
+
+    def infer_label_type(self):
+        label_shape = self.get_label_shape()
```
Comment (noamzbr) [inline deepchecks/vision/dataset.py:82]: feel that given the label_transformer, we should get rid of this. It's the user's responsibility from now on to transform the label to a fitting format. 

It is our responsibility to check that the resulting output matches our requirement, so we should add validation for that instead of this.

File: deepchecks/vision/dataset.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+class TaskType(Enum):
+    """Enum containing supported task types."""
+
+    CLASSIFICATION = 'classification'
+    OBJECT_DETECTION = 'object_detection'
+    SEMANTIC_SEGMENTATION = 'semantic_segmentation'
+
+
+class VisionDataset:
```
Comment (noamzbr) [inline deepchecks/vision/dataset.py:25]: TODO: add docstring. Should elaborate also on required label format after label_transformer

File: deepchecks/vision/datasets/classification/mnist.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import torch
+import torchvision
+
+
+import pathlib
+current_path = pathlib.Path(__file__).parent.resolve()
+model_path = str(current_path).replace('\\', '/') + '/models/mnist.pth'
+
+
+class MNistNet(nn.Module):
```
Comment (noamzbr) [inline deepchecks/vision/datasets/classification/mnist.py:14]: This file doesn't really work due to model_path. Needs refactor. Suggest removing the model for now and re-adding it once it works standalone.

File: deepchecks/vision/datasets/detection/yolo_to_coco.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+              "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl",
+              "banana", "apple", "sandwich", "orange", "broccoli", "carrot",
+              "hot dog", "pizza", "donut", "cake", "chair", "couch",
+              "potted plant", "bed", "dining table", "toilet", "tv", "laptop",
+              "mouse", "remote", "keyboard", "cell phone", "microwave",
+              "oven", "toaster", "sink", "refrigerator", "book", "clock",
+              "vase", "scissors", "teddy bear", "hair drier", "toothbrush")
+
+
+class YoloParser:
```
Comment (noamzbr) [inline deepchecks/vision/datasets/detection/yolo_to_coco.py:29]: Doesn't serve any purpose RN. Should perhaps be kept aside and added later to utils for users wishing to convert coco json format to our accepted format.

File: deepchecks/vision/utils/accuracy_multiclass.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+from ignite.metrics import Metric
```
Comment (noamzbr) [inline deepchecks/vision/utils/accuracy_multiclass.py:1]: File should be removed, or otherwise changed to single-class metric used in example notebook of custom metric.

File: deepchecks/vision/utils/detection_precision_recall.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
+from collections import defaultdict
+
+from ignite.metrics import Metric
+from ignite.metrics.metric import sync_all_reduce, reinit__is_reduced
+import torch
+import numpy as np
+from .iou_utils import compute_ious
+
+
+class DetectionPrecisionRecall(Metric):
```
Comment (noamzbr) [inline deepchecks/vision/utils/detection_precision_recall.py:10]: For now should be renamed to AveragePrecision, because this is what this will display until we refactor it to return many values.

File: deepchecks/vision/utils/iou_utils.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+from collections import defaultdict
```
Comment (noamzbr) [inline deepchecks/vision/utils/iou_utils.py:1]: metric_utils?

preparing for future utils...

File: deepchecks/vision/utils/validation.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import typing as t
+
+from torch import nn
+
+from deepchecks.core import errors
+
+
+def model_type_validation(model: t.Any):
```
Comment (noamzbr) [inline deepchecks/vision/utils/validation.py:8]: not actually needed. We need to validate the model(batch[0]) runs

File: tests/conftest.py [hunk 11]
```diff
@@ -25,6 +25,7 @@
 
 from deepchecks.tabular import Dataset
 
+from .vision import vision_conftest
```
Comment (noamzbr) [inline tests/conftest.py:28]: won't this load all vision stuff when testing the general package? Do we want that?

File: deepchecks/vision/checks/performance/performance_report.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        """
+        return self._performance_report(train_dataset, test_dataset, model)
+
+    def _performance_report(self, train_dataset: VisionDataset, test_dataset: VisionDataset, model):
+        VisionDataset.validate_dataset(train_dataset)
+        VisionDataset.validate_dataset(test_dataset)
+        train_dataset.validate_label()
+        test_dataset.validate_label()
+        train_dataset.validate_shared_label(test_dataset)
+        model_type_validation(model)
```
Comment (noamzbr) [inline deepchecks/vision/checks/performance/performance_report.py:65]: TODO: add validation that self.prediction_extract(model(batch[0])) returns the format we want. Should be in metric utils or something.

File: deepchecks/vision/__init__.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    ModelOnlyBaseCheck
+)
+
+logger = logging.getLogger("deepchecks")
+
+try:
+    import torch
+except ImportError:
+    logger.error("PyTorch is not installed. Please install it in order to use deepchecks.vision")
+
```
Comment (yromanyshyn) [inline deepchecks/vision/__init__.py:18]: we should reraise the `ImportError`, not just log the message, 
also would be better to change the message to something like
`deepchecks was not installed with a set of dependencies to support computer vision functionality, please run pip install deepchecks[vision]`

File: deepchecks/vision/base.py [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+    #     single_scorer_dict = {scorer_name: scorers[scorer_name]}
+    #     return init_validate_scorers(single_scorer_dict, self.model, self.train, class_avg, self.task_type)[0]
+
+
+class Check(BaseCheck):
+    """Base class for all tabular checks."""
+
+    def __init__(self):
+        # pylint: disable=super-init-not-called
+        self._conditions = OrderedDict()
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/vision/base.py:243]: Opened issue #781 for this.

File: deepchecks/vision/dataset.py (hunk 2 above)
Comment (noamzbr) [reply deepchecks/vision/dataset.py:87]: Could be resolved as part of #758

File: deepchecks/vision/__init__.py (hunk 13 above)
Comment (noamzbr) [reply deepchecks/vision/__init__.py:29]: @ItayGabbay what's your take? The message @yromanyshyn isn't 100% accurate as installing deepchecks[vision] won't install pytorch (user must do that on his own), but it's a legit option to just raise an error (as nothing in the package would work if he tries it later on).

File: deepchecks/vision/base.py [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+# Copyright (C) 2021 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Module for base tabular abstractions."""
```
Comment (ItayGabbay) [inline deepchecks/vision/base.py:11]:
```suggestion
"""Module for base vision abstractions."""
```

File: deepchecks/vision/checks/performance/performance_report.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+# Copyright (C) 2021 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+# pylint: disable=all
```
Comment (ItayGabbay) [inline deepchecks/vision/checks/performance/performance_report.py:11]: Let's delete this @nirhutnik

File: deepchecks/vision/__init__.py (hunk 13 above)
Comment (ItayGabbay) [reply deepchecks/vision/__init__.py:29]: I'll change it to ImportError

**Review by noamzbr — APPROVED** [review]
